### ISSUE TITLE
[Megatron-FSDP] Gradients doubled every microbatch (2^N amplification) when a grad buffer's DP group has size 1 (EP=DP or DP=1) with optim_grads_params

### ISSUE THREAD

**Issue by XiongFenghhh** [NVIDIA/Megatron-LM#6660]
**Describe the bug**

@NVIDIA/mcore-oncall

With `use_megatron_fsdp=True` and `data_parallel_sharding_strategy='optim_grads_params'`,
any gradient buffer whose data-parallel group degenerates to **a single rank** gets its
content **doubled on every microbatch** during gradient accumulation:

```
buf_k = 2 * buf_{k-1} + g_k      # per microbatch
buf_N ≈ 2^N * g                  # after N microbatches
```

This happens for **expert gradients when EP = DP** (expert-DP group size = DP/EP = 1),
and for **all gradients when DP = 1** (single GPU).

Because the per-microbatch factor is exactly 2, the corruption is invisible at small
microbatch counts (values merely look "large", e.g. ×2^16 ≈ 6.5e4 at 16 microbatches,
which clip-grad then silently rescales away), and only surfaces at high counts:
at 128 microbatches/rank the expert grad absmax reaches ~3.5e34, the fp32 L2-norm
computation overflows, `grad norm = inf`, and clip_grad_norm multiplies **all**
gradients by ~0 — the run "trains" with zeroed gradients while
`number of skipped iterations = 0`.

**Steps/Code to reproduce bug**

Minimal config-level repro (any small MoE model, e.g. 4 layers, 16–144 experts):

- 2 GPUs, TP=1, PP=1, **EP=2** (=> expert-DP group size 1)
- `use_megatron_fsdp=True`, `use_distributed_optimizer=True`,
  `data_parallel_sharding_strategy='optim_grads_params'`, bf16 with fp32 main grads
  (`grad_reduce_in_fp32=True`)
- Gradient accumulation with N microbatches/rank (e.g. mbs=1, GBS=256 on 2 GPUs → N=128)

Observed: `grad norm: inf` at iteration 1, expert-only. Ablations (same seed):
N=16/rank → grad norm ~1e4 (looks fine, is actually ×2^16); N=32 on 1 GPU → ~3e9;
N=128/rank → inf. Norm magnitude scales as 2^N, not linearly with batch size.

To watch the mechanism directly, wrap `schedules.backward_step` and read the expert
grad buffer after every microbatch (absmax doubles every step, from 1.5e-4 upward,
strictly ×2.0):

```
mb#3: 1.575e-03   mb#8: 5.222e-02   mb#13: 1.673e+00   mb#18: 5.354e+01
mb#4: 3.206e-03   mb#9: 1.045e-01   mb#14: 3.346e+00   mb#19: 1.071e+02
mb#5: 6.475e-03   mb#10: 2.090e-01  mb#15: 6.692e+00   mb#20: 2.141e+02
mb#6: 1.299e-02   mb#11: 4.182e-01  mb#16: 1.338e+01   mb#21: 4.283e+02
mb#7: 2.609e-02   mb#12: 8.364e-01  mb#17: 2.677e+01   ...
```

Extrapolation to N=128: 428.3 × 2^106 ≈ 3.5e34, matching the absmax observed at
clip time (3.47e34) exactly.

**Expected behavior**

Accumulated gradient = sum of per-microbatch gradients (linear growth in N).
With EP=2/DP=2 and 128 microbatches, expert grad absmax should be O(1), not ~1e34.

**Additional context**

- Reproduced on core_v0.17.0 (the implicated code is identical in core_v0.17.1).
- The corruption is **silent** in realistic regimes:
  - at N ≤ ~64 microbatches/rank nothing overflows; clip_grad=1.0 uniformly rescales
    the (directionally correct, hugely inflated) expert gradients — training appears
    normal but expert updates are wrong;
  - with clip disabled the run even converges superficially, because `2*buf+g`
    preserves gradient *direction* and Adam normalizes per-element magnitude;
  - `nonfinite=0` throughout (2^k scaling never produces NaN/Inf in the buffer
    itself; only the norm's square overflows fp32).
- Implicated code region (`megatron/core/distributed/fsdp/src/megatron_fsdp/`):
  - `megatron_fsdp.py:668-678` — `grad_reduce_every_bprop=True` for
    `optim_grads_params`/`optim_grads` triggers `reduce_gradients()` on **every**
    microbatch (the "per microbatch" in 2^N);
  - `param_and_grad_buffer.py:2302` — `is_data_distributed = ... and
    dp_group.size() > 1` is where EDP=1 expert groups diverge from dense params;
  - `param_and_grad_buffer.py:3566-3682` `_bucket_group_gradient_reduce()` —
    the reduce + accumulation closure (`local_grad += reduced_grad` at :3682).
    Note the existing comment at :3660-3665 already anticipates buffer aliasing
    ("the reduce-scatter result has already been installed into the gradient
    buffer"), but a degenerate 1-rank group appears to hit an equivalent
    self-add path. In our config grad_comm_dtype == main_grads_dtype == fp32,
    so the `custom_grad_comm_dtype` branch is *not* taken; the exact self-add
    line is therefore not conclusively pinned — but the empirical 2^N recurrence
    is unambiguous, and the trigger region is confined to per-microbatch reduce
    + accumulation for size-1 DP groups.
- Workarounds that avoid the bug (verified by mechanism; EP<DP verified empirically):
  keep **EP < DP** (EDP ≥ 2); use classic DDP + distributed optimizer instead of
  Megatron-FSDP; or `data_parallel_sharding_strategy='optim'` (reduce once per
  iteration). No config workaround exists for DP=1.
- Suggested fix direction: skip the accumulation closure (or the whole per-microbatch
  reduce) for size-1 DP groups, or gate the `+=` closure on the reduced shard not
  aliasing the accumulation buffer.

**Comment by wujingyue** [comment]
Thanks for the report. Could you share a minimal runnable repro?

**Comment by XiongFenghhh** [comment]
> Thanks for the report. Could you share a minimal runnable repro?

our experiment is conduct on a two-gpu environment

torchrun --nproc_per_node 2 pretrain_gpt.py \
  --num-layers 4 --hidden-size 1536 --ffn-hidden-size 1536 \
  --num-attention-heads 16 --group-query-attention --num-query-groups 2 \
  --kv-channels 96 --qk-layernorm --normalization RMSNorm --swiglu \
  --position-embedding-type rope --rotary-base 1000000 \
  --num-experts 144 --moe-ffn-hidden-size 832 --moe-router-topk 8 \
  --moe-shared-expert-intermediate-size 1536 --moe-router-dtype fp32 \
  --moe-token-dispatcher-type alltoall --moe-expert-capacity-factor -1 \
  --untie-embeddings-and-output-weights --disable-bias-linear \
  --seq-length 8192 --max-position-embeddings 8192 \
  --expert-model-parallel-size 2 \
  --use-megatron-fsdp --use-distributed-optimizer \
  --data-parallel-sharding-strategy optim_grads_params \
  --ckpt-format fsdp_dtensor --no-save-optim \
  --bf16 --attention-backend flash --cross-entropy-loss-fusion \
  --no-gradient-accumulation-fusion \
  --micro-batch-size 1 --global-batch-size 256 \
  --train-iters 10 --lr 1e-4 --min-lr 1e-5 \
  --lr-decay-style constant --lr-warmup-iters 0 \
  --mock-data --tokenizer-type NullTokenizer --vocab-size 248320 \
  --log-interval 1

**Comment by wujingyue** [comment]
Thanks for the repro command.

For reference, I tested this on commit `e7b92efadb527c4ad01c2eb1a1d64692818f1b8d` (the `main` tip at the time of testing). To fit the available GPUs, I reduced the vocabulary size to 32000 and added `--eval-interval 10000` for mock-data initialization. The EP/DP topology, 8K sequence length, `optim_grads_params` sharding, disabled gradient-accumulation fusion, and 128 microbatches per rank were unchanged.

```bash
torchrun --nproc_per_node 2 pretrain_gpt.py \
  --num-layers 4 --hidden-size 1536 --ffn-hidden-size 1536 \
  --num-attention-heads 16 --group-query-attention --num-query-groups 2 \
  --kv-channels 96 --qk-layernorm --normalization RMSNorm --swiglu \
  --position-embedding-type rope --rotary-base 1000000 \
  --num-experts 144 --moe-ffn-hidden-size 832 --moe-router-topk 8 \
  --moe-shared-expert-intermediate-size 1536 --moe-router-dtype fp32 \
  --moe-token-dispatcher-type alltoall --moe-expert-capacity-factor -1 \
  --untie-embeddings-and-output-weights --disable-bias-linear \
  --seq-length 8192 --max-position-embeddings 8192 \
  --expert-model-parallel-size 2 \
  --use-megatron-fsdp --use-distributed-optimizer \
  --data-parallel-sharding-strategy optim_grads_params \
  --ckpt-format fsdp_dtensor --no-save-optim \
  --bf16 --attention-backend flash --cross-entropy-loss-fusion \
  --no-gradient-accumulation-fusion \
  --micro-batch-size 1 --global-batch-size 256 \
  --train-iters 10 --lr 1e-4 --min-lr 1e-5 \
  --lr-decay-style constant --lr-warmup-iters 0 \
  --mock-data --tokenizer-type NullTokenizer --vocab-size 32000 \
  --log-interval 1 --eval-interval 10000
```

The 10-iteration run completed with finite gradient norms throughout: `2.006, 4.754, 4.722, 7.307, 7.972, 3.593, 3.623, 3.158, 2.405, 1.948`. These are the standard per-iteration `grad norm` fields emitted by `--log-interval 1`. All iterations reported zero skipped and zero NaN iterations. I could not reproduce the reported $2^N$ gradient growth.

The issue reports `core_v0.17.0` and `core_v0.17.1`; the test above was run at the commit noted above.

**Comment by XiongFenghhh** [comment]
> Thanks for the repro command.
> 
> For reference, I tested this on commit `e7b92efadb527c4ad01c2eb1a1d64692818f1b8d` (the `main` tip at the time of testing). To fit the available GPUs, I reduced the vocabulary size to 32000 and added `--eval-interval 10000` for mock-data initialization. The EP/DP topology, 8K sequence length, `optim_grads_params` sharding, disabled gradient-accumulation fusion, and 128 microbatches per rank were unchanged.
> 
> torchrun --nproc_per_node 2 pretrain_gpt.py \
>   --num-layers 4 --hidden-size 1536 --ffn-hidden-size 1536 \
>   --num-attention-heads 16 --group-query-attention --num-query-groups 2 \
>   --kv-channels 96 --qk-layernorm --normalization RMSNorm --swiglu \
>   --position-embedding-type rope --rotary-base 1000000 \
>   --num-experts 144 --moe-ffn-hidden-size 832 --moe-router-topk 8 \
>   --moe-shared-expert-intermediate-size 1536 --moe-router-dtype fp32 \
>   --moe-token-dispatcher-type alltoall --moe-expert-capacity-factor -1 \
>   --untie-embeddings-and-output-weights --disable-bias-linear \
>   --seq-length 8192 --max-position-embeddings 8192 \
>   --expert-model-parallel-size 2 \
>   --use-megatron-fsdp --use-distributed-optimizer \
>   --data-parallel-sharding-strategy optim_grads_params \
>   --ckpt-format fsdp_dtensor --no-save-optim \
>   --bf16 --attention-backend flash --cross-entropy-loss-fusion \
>   --no-gradient-accumulation-fusion \
>   --micro-batch-size 1 --global-batch-size 256 \
>   --train-iters 10 --lr 1e-4 --min-lr 1e-5 \
>   --lr-decay-style constant --lr-warmup-iters 0 \
>   --mock-data --tokenizer-type NullTokenizer --vocab-size 32000 \
>   --log-interval 1 --eval-interval 10000
> 
> The 10-iteration run completed with finite gradient norms throughout: `2.006, 4.754, 4.722, 7.307, 7.972, 3.593, 3.623, 3.158, 2.405, 1.948`. These are the standard per-iteration `grad norm` fields emitted by `--log-interval 1`. All iterations reported zero skipped and zero NaN iterations. I could not reproduce the reported 2 N gradient growth.
> 
> The issue reports `core_v0.17.0` and `core_v0.17.1`; the test above was run at the commit noted above.

We upgraded our Megatron to 0.18.2 and the issue we encountered disappeared. I think the problem was introduced by the Megatron bundled with slime, which turned out not to be 0.17. Anyway, thank you for your patience.
